# Create and run a local RAG pipeline from scratch

The goal of this notebook is to build a RAG (Retrieval Augmented Generation) pipeline from scratch and have it run on a local GPU.

Specifically, we'd like to be able to open a PDF file, ask questions (queries) of it and have them answered by a Large Language Model (LLM).

There are frameworks that replicate this kind of workflow, including [LlamaIndex](https://www.llamaindex.ai/) and [LangChain](https://www.langchain.com/), however, the goal of building from scratch is to be able to inspect and customize all the parts.

## What is RAG?

RAG stands for Retrieval Augmented Generation.

It was introduced in the paper [*Retrieval-Augmented Generation for Knowledge-Intensive NLP Tasks*](https://arxiv.org/abs/2005.11401).

Each step can be roughly broken down to:

* **Retrieval** - Seeking relevant information from a source given a query. For example, getting relevant passages of Wikipedia text from a database given a question.
* **Augmented** - Using the relevant retrieved information to modify an input to a generative model (e.g. an LLM).
* **Generation** - Generating an output given an input. For example, in the case of an LLM, generating a passage of text given an input prompt.

## Why RAG?

The main goal of RAG is to improve the generation outptus of LLMs.

Two primary improvements can be seen as:
1. **Preventing hallucinations** - LLMs are incredible but they are prone to potential hallucination, as in, generating something that *looks* correct but isn't. RAG pipelines can help LLMs generate more factual outputs by providing them with factual (retrieved) inputs. And even if the generated answer from a RAG pipeline doesn't seem correct, because of retrieval, you also have access to the sources where it came from.
2. **Work with custom data** - Many base LLMs are trained with internet-scale text data. This means they have a great ability to model language, however, they often lack specific knowledge. RAG systems can provide LLMs with domain-specific data such as medical information or company documentation and thus customized their outputs to suit specific use cases.

RAG can also be a much quicker solution to implement than fine-tuning an LLM on specific data. 


## What kind of problems can RAG be used for?

RAG can help anywhere there is a specific set of information that an LLM may not have in its training data (e.g. anything not publicly accessible on the internet).

For example you could use RAG for:
* **Customer support Q&A chat** - By treating your existing customer support documentation as a resource, when a customer asks a question, you could have a system retrieve relevant documentation snippets and then have an LLM craft those snippets into an answer. Think of this as a "chatbot for your documentation". Klarna, a large financial company, [uses a system like this](https://www.klarna.com/international/press/klarna-ai-assistant-handles-two-thirds-of-customer-service-chats-in-its-first-month/) to save $40M per year on customer support costs.
* **Email chain analysis** - Let's say you're an insurance company with long threads of emails between customers and insurance agents. Instead of searching through each individual email, you could retrieve relevant passages and have an LLM create strucutred outputs of insurance claims.
* **Company internal documentation chat** - If you've worked at a large company, you know how hard it can be to get an answer sometimes. Why not let a RAG system index your company information and have an LLM answer questions you may have? The benefit of RAG is that you will have references to resources to learn more if the LLM answer doesn't suffice.
* **Textbook Q&A** - Let's say you're studying for your exams and constantly flicking through a large textbook looking for answers to your quesitons. RAG can help provide answers as well as references to learn more.

All of these have the common theme of retrieving relevant resources and then presenting them in an understandable way using an LLM.

From this angle, you can consider an LLM a calculator for words.


## Why local?

Privacy, speed, cost.

Running locally means you use your own hardware.

From a privacy standpoint, this means you don't have send potentially sensitive data to an API.

From a speed standpoint, it means you won't necessarily have to wait for an API queue or downtime, if your hardware is running, the pipeline can run.

And from a cost standpoint, running on your own hardware often has a heavier starting cost but little to no costs after that.

Performance wise, LLM APIs may still perform better than an open-source model running locally on general tasks but there are more and more examples appearing of smaller, focused models outperforming larger models. 

## Key terms

| Term | Description |
| ----- | ----- | 
| **Token** | A sub-word piece of text. For example, "hello, world!" could be split into ["hello", ",", "world", "!"]. A token can be a whole word, part of a word or group of punctuation characters. 1 token ~= 4 characters in English, 100 tokens ~= 75 words.<br> Text gets broken into tokens before being passed to an LLM. |
| **Embedding** | A learned numerical representation of a piece of data. For example, a sentence of text could be represented by a vector with<br> 768 values. Similar pieces of text (in meaning) will ideally have similar values. |
| **Embedding model** | A model designed to accept input data and output a numerical representation. For example, a text embedding model may take in 384 tokens of text and turn it into a vector of size 768. An embedding model can and often is different to an LLM model. |
| **Similarity search/vector search** | Similarity search/vector search aims to find two vectors which are close together in high-demensional space. For example, two pieces of similar text passed through an embedding model should have a high similarity score, whereas two pieces of text about different topics will have a lower similarity score. Common similarity score measures are dot product and cosine similarity. |
| **Large Language Model (LLM)** | A model which has been trained to numerically represent the patterns in text. A generative LLM will continue a sequence when given a sequence. For example, given a sequence of the text "hello, world!", a genertive LLM may produce "we're going to build a RAG pipeline today!". This generation will be highly dependant on the training data and prompt. |
| **LLM context window** | The number of tokens a LLM can accept as input. For example, as of March 2024, GPT-4 has a default context window of 32k tokens (about 96 pages of text) but can go up to 128k if needed. A recent open-source LLM from Google, Gemma (March 2024) has a context window of 8,192 tokens (about 24 pages of text). A higher context window means an LLM can accept more relevant information to assist with a query. For example, in a RAG pipeline, if a model has a larger context window, it can accept more reference items from the retrieval system to aid with its generation. |
| **Prompt** | A common term for describing the input to a generative LLM. The idea of "[prompt engineering](https://en.wikipedia.org/wiki/Prompt_engineering)" is to structure a text-based (or potentially image-based as well) input to a generative LLM in a specific way so that the generated output is ideal. This technique is possible because of a LLMs capacity for in-context learning, as in, it is able to use its representation of language to breakdown the prompt and recognize what a suitable output may be (note: the output of LLMs is probable, so terms like "may output" are used). | 




 ## What we're going to build

We're going to build RAG pipeline which enables us to chat with a PDF document, specifically an open-source [nutrition textbook](https://pressbooks.oer.hawaii.edu/humannutrition2/), ~1200 pages long.

You could call our project NutriChat!

We'll write the code to:
1. Open a PDF document (you could use almost any PDF here).
2. Format the text of the PDF textbook ready for an embedding model (this process is known as text splitting/chunking).
3. Embed all of the chunks of text in the textbook and turn them into numerical representation which we can store for later.
4. Build a retrieval system that uses vector search to find relevant chunks of text based on a query.
5. Create a prompt that incorporates the retrieved pieces of text.
6. Generate an answer to a query based on passages from the textbook.

The above steps can broken down into two major sections:
1. Document preprocessing/embedding creation (steps 1-3).
2. Search and answer (steps 4-6).

And that's the structure we'll follow.

It's similar to the workflow outlined on the NVIDIA blog which [details a local RAG pipeline](https://developer.nvidia.com/blog/rag-101-demystifying-retrieval-augmented-generation-pipelines/).

<img src="https://github.com/mrdbourke/simple-local-rag/blob/main/images/simple-local-rag-workflow-flowchart.png?raw=true" alt="flowchart of a local RAG workflow" />

## Requirements and setup

* Local NVIDIA GPU (I used a NVIDIA RTX 4090 on a Windows 11 machine) or Google Colab with access to a GPU.
* Environment setup 
* Data source (for example, a PDF). 
* Internet connection (to download the models, but once you have them, it'll run offline).

## 1. Document/Text Processing and Embedding Creation

Ingredients:
* PDF document of choice.
* Embedding model of choice.

Steps:
1. Import PDF document.
2. Process text for embedding (e.g. split into chunks of sentences).
3. Embed text chunks with embedding model.
4. Save embeddings to file for later use (embeddings will store on file for many years or until you lose your hard drive).

### Import PDF Document 

This will work with many other kinds of documents.

However, we'll start with PDF since many people have PDFs.

But just keep in mind, text files, email chains, support documentation, articles and more can also work.

We're going to pretend we're nutrition students at the University of Hawai'i, reading through the open-source PDF textbook [*Human Nutrition: 2020 Edition*](https://pressbooks.oer.hawaii.edu/humannutrition2/).

There are several libraries to open PDFs with Python but I found that [PyMuPDF](https://github.com/pymupdf/pymupdf) works quite well in many cases.

First we'll download the PDF if it doesn't exist.

In [1]:
# Download PDF file
import os
import requests

# Get PDF document
pdf_path = "human-nutrition-text.pdf"

# Download PDF if it doesn't already exist
if not os.path.exists(pdf_path):
  print("File doesn't exist, downloading...")

  # The URL of the PDF you want to download
  url = "https://pressbooks.oer.hawaii.edu/humannutrition2/open/download?type=pdf"

  # The local filename to save the downloaded file
  filename = pdf_path

  # Send a GET request to the URL
  response = requests.get(url)

  # Check if the request was successful
  if response.status_code == 200:
      # Open a file in binary write mode and save the content to it
      with open(filename, "wb") as file:
          file.write(response.content)
      print(f"The file has been downloaded and saved as {filename}")
  else:
      print(f"Failed to download the file. Status code: {response.status_code}")
else:
  print(f"File {pdf_path} exists.")

File human-nutrition-text.pdf exists.


PDF acquired!

We can import the pages of our PDF to text by first defining the PDF path and then opening and reading it with PyMuPDF (`import fitz`).

We'll write a small helper function to preprocess the text as it gets read. Note that not all text will be read in the same so keep this in mind for when you prepare your text.

We'll save each page to a dictionary and then append that dictionary to a list for ease of use later.

In [2]:
import fitz
from tqdm.auto import tqdm

def text_formatter(text: str) -> str:
    """Performs minor formatting on text."""
    cleaned_text = text.replace("\n", " ").strip()
    return cleaned_text

# Open PDF and get lines/pages
# Note: this only focuses on text, rather than images/figures etc
def open_and_read_pdf(pdf_path: str) -> list[dict]:
    """
    Opens a PDF file, reads its text content page by page, and collects statistics.

    Parameters:
        pdf_path (str): The file path to the PDF document to be opened and read.

    Returns:
        list[dict]: A list of dictionaries, each containing the page number
        (adjusted), character count, word count, sentence count, token count, and the extracted text
        for each page.
    """
    doc = fitz.open(pdf_path)  # open a document
    pages_and_texts = []
    for page_number, page in tqdm(enumerate(doc)):  # iterate the document pages
        text = page.get_text()  # get plain text encoded as UTF-8
        text = text_formatter(text)
        pages_and_texts.append({"page_number": page_number - 41,  # adjust page numbers since our PDF starts on page 42
                                "page_char_count": len(text),
                                "page_word_count": len(text.split(" ")),
                                "page_sentence_count_raw": len(text.split(". ")),
                                "page_token_count": len(text) / 4,  # 1 token = ~4 chars, see: https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
                                "text": text})
    return pages_and_texts

pages_and_texts = open_and_read_pdf(pdf_path=pdf_path)
pages_and_texts[:2]

0it [00:00, ?it/s]

[{'page_number': -41,
  'page_char_count': 29,
  'page_word_count': 4,
  'page_sentence_count_raw': 1,
  'page_token_count': 7.25,
  'text': 'Human Nutrition: 2020 Edition'},
 {'page_number': -40,
  'page_char_count': 0,
  'page_word_count': 1,
  'page_sentence_count_raw': 1,
  'page_token_count': 0.0,
  'text': ''}]

Now let's get a random sample of the pages.

In [3]:
import random
random.sample(pages_and_texts, k=3)

[{'page_number': 876,
  'page_char_count': 815,
  'page_word_count': 133,
  'page_sentence_count_raw': 10,
  'page_token_count': 203.75,
  'text': 'Magnesium (mg)  130.0  Niacin(B3) (mg)  8.0  Phosphorus (mg)  500.0  Riboflavin (B2) (mcg)  600.0  Selenium (mcg)  30.0  Thiamine (B1) (mcg)  600.0  Zinc (mg)  5.0  Source:Institute of Medicine. 2006. Dietary Reference Intakes: The  Essential Guide to Nutrient Requirements. Washington, DC: The  National Academies Press. https://doi.org/10.17226/11537. Accessed  December 10, 2017.  Factors Influencing Intake  A number of factors can influence children’s eating habits and  attitudes toward food. Family environment, societal trends, taste  preferences, and messages in the media all impact the emotions  that children develop in relation to their diet. Television  commercials can entice children to consume sugary products, fatty  fast-foods, excess calories, refined ingredients, and sodium.  876  |  Childhood'},
 {'page_number': 308,
  'page_cha

### Get some stats on the text

Let's perform a rough exploratory data analysis (EDA) to get an idea of the size of the texts (e.g. character counts, word counts etc) we're working with.

The different sizes of texts will be a good indicator into how we should split our texts.

Many embedding models have limits on the size of texts they can ingest, for example, the [`sentence-transformers`](https://www.sbert.net/docs/pretrained_models.html) model [`all-mpnet-base-v2`](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) has an input size of 384 tokens.

This means that the model has been trained in ingest and turn into embeddings texts with 384 tokens (1 token ~= 4 characters ~= 0.75 words).

Texts over 384 tokens which are encoded by this model will be auotmatically reduced to 384 tokens in length, potentially losing some information.

We'll discuss this more in the embedding section.

For now, let's turn our list of dictionaries into a DataFrame and explore it.

In [4]:
import pandas as pd

df = pd.DataFrame(pages_and_texts)
df.head()

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
0          -41               29                4                        1   
1          -40                0                1                        1   
2          -39              320               54                        1   
3          -38              212               32                        1   
4          -37              797              147                        3   

   page_token_count                                               text  
0              7.25                      Human Nutrition: 2020 Edition  
1              0.00                                                     
2             80.00  Human Nutrition: 2020  Edition  UNIVERSITY OF ...  
3             53.00  Human Nutrition: 2020 Edition by University of...  
4            199.25  Contents  Preface  University of Hawai‘i at Mā...

In [5]:
# Get stats
df.describe().round(2)

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
count      1208.00          1208.00          1208.00                  1208.00   
mean        562.50          1148.00           199.50                    10.52   
std         348.86           560.38            95.83                     6.55   
min         -41.00             0.00             1.00                     1.00   
25%         260.75           762.00           134.00                     5.00   
50%         562.50          1231.50           216.00                    10.00   
75%         864.25          1603.50           272.00                    15.00   
max        1166.00          2308.00           430.00                    39.00   

       page_token_count  
count           1208.00  
mean             287.00  
std              140.10  
min                0.00  
25%              190.50  
50%              307.88  
75%              400.88  
max              577.00

Okay, looks like our average token count per page is 287.

For this particular use case, it means we could embed an average whole page with the `all-mpnet-base-v2` model (this model has an input capacity of 384).

### Further text processing (splitting pages into sentences)

The ideal way of processing text before embedding it is still an active area of research.

A simple method I've found helpful is to break the text into chunks of sentences.

As in, chunk a page of text into groups of 5, 7, 10 or more sentences (these values are not set in stone and can be explored).

But we want to follow the workflow of:

`Ingest text -> split it into groups/chunks -> embed the groups/chunks -> use the embeddings`

Some options for splitting text into sentences:

1. Split into sentences with simple rules (e.g. split on ". " with `text = text.split(". ")`, like we did above).
2. Split into sentences with a natural language processing (NLP) library such as [spaCy](https://spacy.io/) or [nltk](https://www.nltk.org/).

Why split into sentences?

* Easier to handle than larger pages of text (especially if pages are densely filled with text).
* Can get specific and find out which group of sentences were used to help within a RAG pipeline.

Let's use spaCy to break our text into sentences since it's likely a bit more robust than just using `text.split(". ")`. 

In [6]:
from spacy.lang.en import English

nlp = English()

# Add a sentencizer pipeline, see https://spacy.io/api/sentencizer/ 
nlp.add_pipe("sentencizer")

# Create a document instance as an example
doc = nlp("This is a sentence. This is another sentence.")
assert len(list(doc.sents)) == 2

# Access the sentences of the document
list(doc.sents)

[This is a sentence., This is another sentence.]

We don't necessarily need to use spaCy, however, it's an open-source library designed to do NLP tasks like this at scale.

So let's run our small sentencizing pipeline on our pages of text.

In [7]:
for item in tqdm(pages_and_texts):
    item["sentences"]=list(nlp(item["text"]).sents)

    # Make sure all sentences are strings
    item["sentences"] = [str(sentence) for sentence in item["sentences"]]
    
    # Count the sentences 
    item["page_sentence_count_spacy"] = len(item["sentences"])

  0%|          | 0/1208 [00:00<?, ?it/s]

In [8]:
# Inspect an example
random.sample(pages_and_texts, k=1)

[{'page_number': 1134,
  'page_char_count': 1648,
  'page_word_count': 264,
  'page_sentence_count_raw': 22,
  'page_token_count': 412.0,
  'text': 'Food Insecurity  Food insecurity is defined as not having adequate access to food  that meets nutritional needs. According to the USDA, about 48.8  million people live in food-insecure households and have reported  multiple indications of food access problems. About sixteen million  of those have “very low food security,” which means one or more  people in the household were hungry at some point over the course  of a year due to the inability to afford enough food. The difference  between low and very low food security is that members of low  insecurity households have reported problems of food access, but  have reported only a few instances of reduced food intake, if any.2  African American and Hispanic households experience food  insecurity at much higher rates than the national average.3  Households with limited resources employ a varie

Wonderful!

Now let's turn out list of dictionaries into a DataFrame and get some stats.

In [9]:
df=pd.DataFrame(pages_and_texts)
df.describe().round(2)

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
count      1208.00          1208.00          1208.00                  1208.00   
mean        562.50          1148.00           199.50                    10.52   
std         348.86           560.38            95.83                     6.55   
min         -41.00             0.00             1.00                     1.00   
25%         260.75           762.00           134.00                     5.00   
50%         562.50          1231.50           216.00                    10.00   
75%         864.25          1603.50           272.00                    15.00   
max        1166.00          2308.00           430.00                    39.00   

       page_token_count  page_sentence_count_spacy  
count           1208.00                    1208.00  
mean             287.00                      10.32  
std              140.10                       6.30  
min                0.00                       0.00  
25%              190.50                       5.00  
50%              307.88                      10.00  
75%              400.88                      15.00  
max              577.00                      28.00

For our set of text, it looks like our raw sentence count (e.g. splitting on `". "`) is quite close to what spaCy came up with.

Now we've got our text split into sentences, how about we group those sentences?

### Chunking our sentences together

Let's take a step to break down our list of sentences/text into smaller chunks.

As you might've guessed, this process is referred to as **chunking**.

Why do we do this?

1. Easier to manage similar sized chunks of text.
2. Don't overload the embedding models capacity for tokens (e.g. if an embedding model has a capacity of 384 tokens, there could be information loss if you try to embed a sequence of 400+ tokens).
3. Our LLM context window (the amount of tokens an LLM can take in) may be limited and requires compute power so we want to make sure we're using it as well as possible.

Something to note is that there are many different ways emerging for creating chunks of information/text.

For now, we're going to keep it simple and break our pages of sentences into groups of 10 (this number is arbitrary and can be changed, I just picked it because it seemed to line up well with our embedding model capacity of 384).

On average each of our pages has 10 sentences.

And an average total of 287 tokens per page.

So our groups of 10 sentences will also be ~287 tokens long.

This gives us plenty of room for the text to embedded by our `all-mpnet-base-v2` model (it has a capacity of 384 tokens).

To split our groups of sentences into chunks of 10 or less, let's create a function which accepts a list as input and recursively breaks into down into sublists of a specified size.

In [10]:
# Define split size to turn groups of sentences into chunks
num_sentence_chunk_size = 10 

# Create a function that recursively splits a list into desired sizes
def split_list(input_list: list, 
               slice_size: int) -> list[list[str]]:
    """
    Splits the input_list into sublists of size slice_size (or as close as possible).

    For example, a list of 17 sentences would be split into two lists of [[10], [7]]
    """
    return [input_list[i:i + slice_size] for i in range(0, len(input_list), slice_size)]

# Loop through pages and texts and split sentences into chunks
for item in tqdm(pages_and_texts):
    item["sentence_chunks"] = split_list(input_list=item["sentences"],
                                         slice_size=num_sentence_chunk_size)
    item["num_chunks"] = len(item["sentence_chunks"])

  0%|          | 0/1208 [00:00<?, ?it/s]

In [11]:
# Sample an example from the group (note: many samples have only 1 chunk as they have <=10 sentences total)
random.sample(pages_and_texts, k=1)

[{'page_number': 888,
  'page_char_count': 744,
  'page_word_count': 123,
  'page_sentence_count_raw': 5,
  'page_token_count': 186.0,
  'text': 'Learning Activities  Technology Note: The second edition of the Human  Nutrition Open Educational Resource (OER) textbook  features interactive learning activities.  These activities are  available in the web-based textbook and not available in the  downloadable versions (EPUB, Digital PDF, Print_PDF, or  Open Document).  Learning activities may be used across various mobile  devices, however, for the best user experience it is strongly  recommended that users complete these activities using a  desktop or laptop computer and in Google Chrome.    An interactive or media element has been  excluded from this version of the text. You can  view it online here:  http://pressbooks.oer.hawaii.edu/ humannutrition2/?p=469    888  |  Adolescence',
  'sentences': ['Learning Activities  Technology Note: The second edition of the Human  Nutrition Open Educ

In [12]:
# Create a DataFrame to get stats
df = pd.DataFrame(pages_and_texts)
df.describe().round(2)

page_number  page_char_count  page_word_count  page_sentence_count_raw  \
count      1208.00          1208.00          1208.00                  1208.00   
mean        562.50          1148.00           199.50                    10.52   
std         348.86           560.38            95.83                     6.55   
min         -41.00             0.00             1.00                     1.00   
25%         260.75           762.00           134.00                     5.00   
50%         562.50          1231.50           216.00                    10.00   
75%         864.25          1603.50           272.00                    15.00   
max        1166.00          2308.00           430.00                    39.00   

       page_token_count  page_sentence_count_spacy  num_chunks  
count           1208.00                    1208.00     1208.00  
mean             287.00                      10.32        1.53  
std              140.10                       6.30        0.64  
min                0.00                       0.00        0.00  
25%              190.50                       5.00        1.00  
50%              307.88                      10.00        1.00  
75%              400.88                      15.00        2.00  
max              577.00                      28.00        3.00

Note how the average number of chunks is around 1.5, this is expected since many of our pages only contain an average of 10 sentences.

### Splitting each chunk into its own item

We'd like to embed each chunk of sentences into its own numerical representation.

So to keep things clean, let's create a new list of dictionaries each containing a single chunk of sentences with relative information such as page number as well statistics about each chunk.

In [13]:
import re

# Split each chunk into its own item
pages_and_chunks = []
for item in tqdm(pages_and_texts):
    for sentence_chunk in item["sentence_chunks"]:
        chunk_dict = {}
        chunk_dict["page_number"] = item["page_number"]
        
        # Join the sentences together into a paragraph-like structure, aka a chunk (so they are a single string)
        joined_sentence_chunk = "".join(sentence_chunk).replace("  ", " ").strip()
        joined_sentence_chunk = re.sub(r'\.([A-Z])', r'. \1', joined_sentence_chunk) # ".A" -> ". A" for any full-stop/capital letter combo 
        chunk_dict["sentence_chunk"] = joined_sentence_chunk

        # Get stats about the chunk
        chunk_dict["chunk_char_count"] = len(joined_sentence_chunk)
        chunk_dict["chunk_word_count"] = len([word for word in joined_sentence_chunk.split(" ")])
        chunk_dict["chunk_token_count"] = len(joined_sentence_chunk) / 4 # 1 token = ~4 characters
        
        pages_and_chunks.append(chunk_dict)

# How many chunks do we have?
len(pages_and_chunks)

  0%|          | 0/1208 [00:00<?, ?it/s]

1843

In [14]:
random.sample(pages_and_chunks, k=1)

[{'page_number': 518,
  'sentence_chunk': 'Image by Allison Calabrese / CC BY 4.0 One major difference between fat-soluble vitamins and water- soluble vitamins is the way they are absorbed in the body. Vitamins are absorbed primarily in the small intestine and their bioavailability is dependent on the food composition of the diet. Fat-soluble vitamins are absorbed along with dietary fat. Therefore, if a meal is very low in fat, the absorption of the fat-soluble vitamins will be impaired. Once fat-soluble vitamins have been absorbed in the small intestine, they are packaged and incorporated into chylomicrons along with other fatty acids and transported in the lymphatic system to the liver. Water-soluble vitamins on the other hand are absorbed in the small intestine but are transported to the liver through blood vessels. Figure 9.2 “Absorption of Fat-Soluble and Water-Soluble Vitamins  518 | Introduction',
  'chunk_char_count': 872,
  'chunk_word_count': 137,
  'chunk_token_count': 218.0

Excellent!

Now we've broken our whole textbook into chunks of 10 sentences or less as well as the page number they came from.

This means we could reference a chunk of text and know its source.

Let's get some stats about our chunks.

In [15]:
# Create a DataFrame to get stats
df = pd.DataFrame(pages_and_chunks)
df.describe().round(2)

page_number  chunk_char_count  chunk_word_count  chunk_token_count
count      1843.00           1843.00           1843.00            1843.00
mean        583.38            734.10            112.74             183.52
std         347.79            447.51             71.24             111.88
min         -41.00             12.00              3.00               3.00
25%         280.50            315.00             45.00              78.75
50%         586.00            745.00            115.00             186.25
75%         890.00           1118.00            173.00             279.50
max        1166.00           1830.00            297.00             457.50

Hmm looks like some of our chunks have quite a low token count.

How about we check for samples with less than 30 tokens (about the length of a sentence) and see if they are worth keeping?

In [16]:
# Show random chunks with under 30 tokens in length
min_token_length = 30
for row in df[df["chunk_token_count"] <= min_token_length].sample(5).iterrows():
    print(f'Chunk token count: {row[1]["chunk_token_count"]} | Text: {row[1]["sentence_chunk"]}')

Chunk token count: 3.75 | Text: 806 | Pregnancy
Chunk token count: 9.25 | Text: 490 | Factors Affecting Energy Intake
Chunk token count: 24.25 | Text: There are several lecithin supplements on the market Nonessential and Essential Fatty Acids | 315
Chunk token count: 20.75 | Text: Centers for Disease Control and Prevention.http://www.cdc.gov/nutrition/ Iron | 661
Chunk token count: 10.5 | Text: Consequences of Deficiency or Excess | 205


Looks like many of these are headers and footers of different pages.

They don't seem to offer too much information.

Let's filter our DataFrame/list of dictionaries to only include chunks with over 30 tokens in length.

In [17]:
pages_and_chunks_over_min_token_len = df[df["chunk_token_count"] > min_token_length].to_dict(orient="records")
pages_and_chunks_over_min_token_len[:2]

[{'page_number': -39,
  'sentence_chunk': 'Human Nutrition: 2020 Edition UNIVERSITY OF HAWAI‘I AT MĀNOA FOOD SCIENCE AND HUMAN NUTRITION PROGRAM ALAN TITCHENAL, SKYLAR HARA, NOEMI ARCEO CAACBAY, WILLIAM MEINKE-LAU, YA-YUN YANG, MARIE KAINOA FIALKOWSKI REVILLA, JENNIFER DRAPER, GEMADY LANGFELDER, CHERYL GIBBY, CHYNA NICOLE CHUN, AND ALLISON CALABRESE',
  'chunk_char_count': 308,
  'chunk_word_count': 42,
  'chunk_token_count': 77.0},
 {'page_number': -38,
  'sentence_chunk': 'Human Nutrition: 2020 Edition by University of Hawai‘i at Mānoa Food Science and Human Nutrition Program is licensed under a Creative Commons Attribution 4.0 International License, except where otherwise noted.',
  'chunk_char_count': 210,
  'chunk_word_count': 30,
  'chunk_token_count': 52.5}]

Smaller chunks filtered!

Time to embed our chunks of text!

### Embedding our text chunks

While humans understand text, machines understand numbers best.

An [embedding](https://vickiboykis.com/what_are_embeddings/index.html) is a broad concept.

The most powerful thing about modern embeddings is that they are *learned* representations.

Meaning rather than directly mapping words/tokens/characters to numbers directly (e.g. `{"a": 0, "b": 1, "c": 3...}`), the numerical representation of tokens is learned by going through large corpuses of text and figuring out how different tokens relate to each other.

Ideally, embeddings of text will mean that similar meaning texts have similar numerical representation.


Our goal is to turn each of our chunks into a numerical representation (an embedding vector, where a vector is a sequence of numbers arranged in order).

Once our text samples are in embedding vectors, us humans will no longer be able to understand them.

However, we don't need to.

The embedding vectors are for our computers to understand.

We'll use our computers to find patterns in the embeddings and then we can use their text mappings to further our understanding.

Enough talking, how about we import a text embedding model and see what an embedding looks like.

To do so, we'll use the [`sentence-transformers`](https://www.sbert.net/docs/installation.html) library which contains many pre-trained embedding models.

Specifically, we'll get the `all-mpnet-base-v2` model (you can see the model's intended use on the [Hugging Face model card](https://huggingface.co/sentence-transformers/all-mpnet-base-v2#intended-uses)).

In [18]:
from sentence_transformers import SentenceTransformer

embedding_model=SentenceTransformer(model_name_or_path='all-mpnet-base-v2', device='cpu')

#Create a list of Sentences
sentences=["I like football","I like basketball", "Kobe is a great player"]

#Sentences are encoded/embedded by calling model.encode()
embeddings=embedding_model.encode(sentences)
embeddings_dict=dict(zip(sentences, embeddings))

#See embeddings
for sentence, embeddings in embeddings_dict.items():
    print(f"Sentence: {sentence}")
    print(f"Embeddings: {embeddings}")
    print("")



Sentence: I like football
Embeddings: [-8.42475444e-02  1.00637212e-01 -3.04262508e-02 -4.04835865e-03
  3.78771499e-02  3.74091119e-02 -9.98117849e-02  2.61343946e-03
  1.68801006e-02 -1.06646698e-02 -9.36578587e-03 -4.31236029e-02
  7.33321533e-03  1.65348276e-02  1.87982209e-02 -7.59599684e-03
 -2.98678111e-02 -2.66341437e-02  1.27094649e-02 -1.01594571e-02
  8.85396637e-03  4.13204506e-02 -1.35907661e-02 -4.72090393e-02
 -1.78263839e-02  4.68517244e-02 -7.70462630e-03 -4.88640293e-02
 -3.99428159e-02  1.04912810e-01 -1.95646044e-02 -3.23913395e-02
 -5.61739877e-02 -5.53603768e-02  1.33881906e-06  1.68700833e-02
 -7.19039282e-03 -2.36500334e-02  6.86780587e-02  2.94983350e-02
  4.66499925e-02  3.60918837e-03 -4.46939059e-02 -4.21173312e-02
 -1.59445629e-02  4.67798449e-02  3.78085896e-02  2.07526819e-03
 -2.71698646e-02  1.42831588e-02  5.78409992e-03 -4.98900842e-03
 -5.02601489e-02 -1.26381991e-02 -4.07727212e-02  1.14967311e-02
  3.48693170e-02  2.09056195e-02  3.40205804e-02  4.

Woah! That's a lot of numbers.

How about we do just once sentence?

In [19]:
single_sentence = "Who is Michael Jordan?"
single_embedding = embedding_model.encode(single_sentence)
print(f"Sentence: {single_sentence}")
print(f"Embedding:\n{single_embedding}")
print(f"Embedding size: {single_embedding.shape}")

Sentence: Who is Michael Jordan?
Embedding:
[ 1.48610771e-03 -1.29408650e-02  4.12214510e-02  1.48524977e-02
  4.33810474e-03  1.77816706e-04  1.77243222e-02 -1.45564685e-02
  3.38754840e-02 -3.48900668e-02  2.58417353e-02  2.15714537e-02
  1.22249816e-02  2.01746691e-02  3.11459322e-02 -1.97694656e-02
  2.88256221e-02 -2.19796542e-02  1.69615056e-02 -6.30995631e-03
 -3.76892760e-02  3.13616637e-03  1.31576005e-02  4.23933007e-02
 -6.79306760e-02 -3.42432857e-02  1.42761124e-02 -1.04357312e-02
 -3.40646617e-02  2.10747235e-02 -5.73627837e-02 -3.49448919e-02
  3.54876113e-03 -5.57608232e-02  1.40041789e-06  8.85181408e-03
  5.87321296e-02 -1.26579041e-02  4.14434522e-02 -3.53717878e-02
 -3.66200916e-02 -4.22541238e-02 -6.18494349e-03  2.17272546e-02
 -1.62388813e-02 -3.43839340e-02  1.98135171e-02 -2.18353979e-02
 -1.34257283e-02  9.53534525e-03  4.17782478e-02 -1.07229194e-02
  3.07925977e-02 -3.29672098e-02 -3.48641463e-02 -1.76475141e-02
  2.64758412e-02  5.72366677e-02 -9.53142047e-

Nice! We've now got a way to numerically represent each of our chunks.

Our embedding has a shape of `(768,)` meaning it's a vector of 768 numbers which represent our text in high-dimensional space, too many for a human to comprehend but machines love high-dimensional space.

How about we add an embedding field to each of our chunk items?

Let's start by trying to create embeddings on the CPU, we'll time it with the `%%time` magic to see how long it takes.

In [20]:
%%time

# the model is on the CPU
embedding_model.to("cpu")

# Embed each chunk one by one
for item in tqdm(pages_and_chunks_over_min_token_len):
    item["embedding"] = embedding_model.encode(item["sentence_chunk"])

  0%|          | 0/1680 [00:00<?, ?it/s]

CPU times: user 11min 38s, sys: 8.17 s, total: 11min 47s
Wall time: 12min 4s


This would take a *really* long time if we had a larger dataset.

In [21]:
%%time

text_chunks=[item['sentence_chunk'] for item in pages_and_chunks_over_min_token_len]
text_chunks[419]

CPU times: user 524 μs, sys: 0 ns, total: 524 μs
Wall time: 570 μs


'often. • Calm your “sweet tooth” by eating fruits, such as berries or an apple. • Replace sugary soft drinks with seltzer water, tea, or a small amount of 100 percent fruit juice added to water or soda water. The Food Industry: Functional Attributes of Carbohydrates and the Use of Sugar Substitutes In the food industry, both fast-releasing and slow-releasing carbohydrates are utilized to give foods a wide spectrum of functional attributes, including increased sweetness, viscosity, bulk, coating ability, solubility, consistency, texture, body, and browning capacity. The differences in chemical structure between the different carbohydrates confer their varied functional uses in foods. Starches, gums, and pectins are used as thickening agents in making jam, cakes, cookies, noodles, canned products, imitation cheeses, and a variety of other foods. Molecular gastronomists use slow- releasing carbohydrates, such as alginate, to give shape and texture to their fascinating food creations. Add

In [22]:
len(text_chunks)

1680

In [23]:
%%time

#Embed all text in batches
text_chunk_embeddings=embedding_model.encode(text_chunks, batch_size=32, convert_to_tensor=True)
text_chunk_embeddings

KeyboardInterrupt: 

### Save embeddings to file

Since creating embeddings can be a timely process (not so much for our case but it can be for more larger datasets), let's turn our `pages_and_chunks_over_min_token_len` list of dictionaries into a DataFrame and save it.

In [24]:
# Save embeddings to file
text_chunks_and_embeddings_df = pd.DataFrame(pages_and_chunks_over_min_token_len)
embeddings_df_save_path = "text_chunks_and_embeddings_df.csv"
text_chunks_and_embeddings_df.to_csv(embeddings_df_save_path, index=False)

And we can make sure it imports nicely by loading it.

In [25]:
# Import saved file and view
text_chunks_and_embedding_df_load = pd.read_csv(embeddings_df_save_path)
text_chunks_and_embedding_df_load.head()

page_number                                     sentence_chunk  \
0          -39  Human Nutrition: 2020 Edition UNIVERSITY OF HA...   
1          -38  Human Nutrition: 2020 Edition by University of...   
2          -37  Contents Preface University of Hawai‘i at Māno...   
3          -36  Lifestyles and Nutrition University of Hawai‘i...   
4          -35  The Cardiovascular System University of Hawai‘...   

   chunk_char_count  chunk_word_count  chunk_token_count  \
0               308                42              77.00   
1               210                30              52.50   
2               766               116             191.50   
3               941               144             235.25   
4               998               152             249.50   

                                           embedding  
0  [ 6.74242601e-02  9.02280062e-02 -5.09550702e-...  
1  [ 5.52156679e-02  5.92137799e-02 -1.66167784e-...  
2  [ 2.79801488e-02  3.39813270e-02 -2.06427258e-...  
3  [ 6.82566613e-02  3.81274521e-02 -8.46859254e-...  
4  [ 3.30264419e-02 -8.49772710e-03  9.57151968e-...